ph5/clients/tests/test_ph5tostationxml.py

```diff
@@ -331,23 +445,41 @@ def setUp(self):
                                  'ph5/test_data/segd/3ch.fcnt')]
         with patch.object(sys, 'argv', testargs):
             segd2ph5.main()
-        self.ph5sxml, self.mng, self.parser = getParser(
-            self.tmpdir, "master.ph5", "CHANNEL")
 
     def tearDown(self):
         self.mng.ph5.close()
-        super(TestPH5toStationXMLParser_resp_load_not_run, self).tearDown()
+        super(TestPH5toStationXMLParser_gen_resp_issue, self).tearDown()
 
-    def test_create_obs_network(self):
-        self.parser.manager.ph5.read_experiment_t()
-        self.parser.experiment_t = self.parser.manager.ph5.Experiment_t['rows']
-        self.parser.add_ph5_stationids()
-        with LogCapture() as log:
-            log.setLevel(logging.ERROR)
-            self.parser.create_obs_network()
-            self.assertEqual(log.records[0].msg,
-                             'All response file names are blank in response '
-                             'table. Check if resp_load has been run.')
+    def test_read_networks_no_resp_filename(self):
+        self.ph5sxml, self.mng, self.parser = getParser(
+            self.tmpdir, "master.ph5", "CHANNEL")
+        with self.assertRaises(ph5tostationxml.PH5toStationXMLError) as contxt:
+            self.parser.read_networks()
+        self.assertEqual(
+            contxt.exception.message,
+            "Response table does not contain any response file names. "
+            "Check if resp_load has been run or if metadatatoph5 input "
+            "contained response information.")
+
+    def test_read_networks_response_duplication(self):
+        # response_t_dup_n_i.kef has n_i= 1,3,6 duplicated
+        testargs = ['delete_table', '-n', 'master.ph5', '-R']
+        with patch.object(sys, 'argv', testargs):
+            nuke_table.main()
+
+        kefpath = os.path.join(self.home,
+                               'ph5/test_data/metadata/response_t_dup_n_i.kef')
+        testargs = ['keftoph5', '-n', 'master.ph5', '-k', kefpath]
+        with patch.object(sys, 'argv', testargs):
+            kef2ph5.main()
+
+        self.ph5sxml, self.mng, self.parser = getParser(
+            self.tmpdir, 'master.ph5', 'NETWORK')
+        with self.assertRaises(ph5tostationxml.PH5toStationXMLError) as contxt:
+            self.parser.read_networks()
+        self.assertEqual(
+            contxt.exception.message,
+            'Response_t n_i(s) duplicated: 1,3,6')
```

Comment:
This is the test that I was trying to create, I previously missed it, and I will delete the test instances that I pushed last night.

The problem is that the master branch will output a stationxml files from a ph5 experiment with duplicated response n_i values in the responses table. It seems like this behavior should likely be mimicked if the flag --stationxml_on_error=True is included in the ph5tostationxml call. I previously suggested that we use a FIFO logical approach and am open to other ideas. @damhuonglan what do you think?

Reply:
response's n_i duplication is checked before each row in array table is checked (Line 686) and error will be thrown if there is any duplication, the same as when there is no response_file_das_a in response_t (Line691).  In case the error is thrown, no stationxml is created at all.

https://github.com/PIC-IRIS/PH5/blob/b7dc1c32047ce6dca221c33a857275cf4a9013c6/ph5/clients/ph5tostationxml.py#L686-L694

I did that because when using response_n_i from array_t to get a row in response_t with the corresponding n_i, n_i must be unique to know which row to get. 

The flag `--stationxml_on_error` is only applied when the errors are logged, but not thrown.

Reply:
The uniqueness problem is an issue, and I understand that if we output stationxml files from ph5 experiments with non-unique n_i value than there is no way to determine if the stationxml document is accurate. The problem is that there are PH5 data in the archive, experiment 18-030, that have this non-unique n_i issue and I am worried that the metadata will get stuck inside these experiments if we don't make some backdoor to access them. I am not sure of the original logic used in ph5tostationxml but we do need some way to output the data that are stored in all of our current ph5 expirments.

If there is no response information in the response table than an error should be thrown and no stationxml should be output.

Reply:
Can you post the response_t of the experiment 18-030 here please?

Reply:
> #
#	Fri Feb 19 22:22:02 2021	ph5 version: 4.1.2
#
#   Table row 1
/Experiment_g/Responses_g/Response_t
	n_i=0
	bit_weight/value_d=7.45058059692e-05
	bit_weight/units_s=mV/count
	gain/units_s=dB
	gain/value_i=12
	response_file_a=
	response_file_das_a=/Experiment_g/Responses_g/ZLAND3C_1000_1_12
	response_file_sensor_a=

#   Table row 2
/Experiment_g/Responses_g/Response_t
	n_i=1
	bit_weight/value_d=0.16985
	bit_weight/units_s=V/count
	gain/units_s=dB
	gain/value_i=24
	response_file_a=
	response_file_das_a=/Experiment_g/Responses_g/RESPGEOD
	response_file_sensor_a=/Experiment_g/Responses_g/RESPGS11D

#   Table row 3
/Experiment_g/Responses_g/Response_t
	n_i=0
	bit_weight/value_d=7.45058059692e-05
	bit_weight/units_s=mV/count
	gain/units_s=dB
	gain/value_i=12
	response_file_a=
	response_file_das_a=
	response_file_sensor_a=

#   Table row 4
/Experiment_g/Responses_g/Response_t
	n_i=1
	bit_weight/value_d=0.16985
	bit_weight/units_s=V/count
	gain/units_s=dB
	gain/value_i=24
	response_file_a=
	response_file_das_a=
	response_file_sensor_a=

Reply:
It is possible there could be other examples of this in the archive... I bet most of them could be resolved though. The current logic could be the correct answer but I just want to make sure we talk about it before moving forward.

Reply:
> It is possible there could be other examples of this in the archive...

16-012 also has a duplicated n_i 

#
#       Thu Dec 17 20:08:42 2020        ph5 version: 4.1.2
#
#   Table row 1
/Experiment_g/Responses_g/Response_t
        n_i=0
        bit_weight/value_d=5.96046447754e-08
        bit_weight/units_s=volts/count
        gain/units_s=
        gain/value_i=32
        response_file_a=
        response_file_das_a=/Experiment_g/Responses_g/rt125_1000_1_32
        response_file_sensor_a=/Experiment_g/Responses_g/GS11V
#   Table row 2
/Experiment_g/Responses_g/Response_t
        n_i=0
        bit_weight/value_d=5.96046447754e-08
        bit_weight/units_s=volts/count
        gain/units_s=
        gain/value_i=32
        response_file_a=
        response_file_das_a=/Experiment_g/Responses_g/rt125_1000_1_32
        response_file_sensor_a=/Experiment_g/Responses_g/L28LB

Reply:
The one that Tim gave won't make any problem with creating stationxml, because the duplicated ones that has no value for response_file_das_a and response_file_sensor_a are coming later
But the one that Holy gave will read the wrong response_file_sensor_a for L28LB because it always read row 1 when querying for n_i=0. Did you create this reponse_t before we fix resp_load?

Reply:
Yes, I originally made this archive in September and October 2019.  The array table for L28LB has response_table_n_i=1, so the xml will not output responses for L28LB at all, but if the array table was not correct the xml will be like you said.  Unfortunately I cannot remember how the mistake happened originally, but how to deal with it is probably more important.

Reply:
I will modify ph5tostationxml to apply flag `--stationxml_on_error` for n_i duplication as Tim ask. In the log message, I will recommend user to run rerun resp_load to fix n_i for the ph5.

Reply:
I think that using the first response with a unique N_I value that appears in the response table is the correct way to handle this issue because we will at least know the specific response that is being output to the stationxml document.